<?xml version="1.0"?>
<robot name="ghost_car" xmlns:xacro="http://ros.org/wiki/xacro">

<xacro:property name="pi" value="3.1415926536" />

<!-- Car Dimensions -->
<xacro:property name="wheelbase" value="0.260" />
<xacro:property name="track_width" value="0.162" />
<xacro:property name="car_height" value="0.08" />

<!-- Colours -->
<material name="blue">
    <color rgba="0 0 0.8 1"/>
</material>
<material name="black">
  <color rgba="0 0 0 1"/>
</material>
<material name="white">
  <color rgba="1 1 1 1"/>
</material>

<!-- Main body -->
<link name="base_footprint">
    <visual>
      <geometry>
        <box size="${wheelbase} ${track_width} ${car_height}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="${0.5*wheelbase} 0 ${0.5*car_height}"/>
      <material name="blue"/>
    </visual>
</link>

<!-- Centre of Gravity -->
<link name="cog"/>
<joint name="cog_joint" type="fixed">
  <origin xyz="${0.5*wheelbase} 0 ${0.5*car_height}" rpy="0 0 0" />
  <parent link="base_footprint"/>
  <child link="cog" />
</joint>

<!-- IMU -->
<!-- Origin is at the centre of board -->
<xacro:property name="imu_length" value="0.02" />
<xacro:property name="imu_width" value="0.03" />
<xacro:property name="imu_height" value="0.004" />
<xacro:property name="imu_x" value="${wheelbase + 0.032}" />
<xacro:property name="imu_y" value="0.0" />
<xacro:property name="imu_z" value="0.038" />
<xacro:property name="imu_yaw" value="0.0" />

<link name="imu">
  <visual>
    <geometry>
      <box size="${imu_length} ${imu_width} ${imu_height}"/>
    </geometry>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <material name="black"/>
  </visual>
</link>
<joint name="imu_mount" type="fixed">
  <parent link="base_footprint"/>
  <child link="imu"/>
  <origin rpy="0 0 ${imu_yaw}" xyz="${imu_x} ${imu_y} ${imu_z}"/>
</joint>

<!-- Camera -->
<xacro:property name="camera_length" value="0.030" />
<xacro:property name="camera_width" value="0.029" />
<xacro:property name="camera_height" value="0.029" />
<xacro:property name="camera_x" value="0.15" />
<xacro:property name="camera_y" value="0" />
<xacro:property name="camera_z" value="0.165" />
<xacro:property name="camera_pitch" value="0.02" />

<link name="camera">
  <visual>
    <geometry>
      <box size="${camera_length} ${camera_width} ${camera_height}"/>
    </geometry>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <material name="white"/>
  </visual>
</link>
<joint name="camera_mount" type="fixed">
  <parent link="base_footprint"/>
  <child link="camera"/>
  <origin rpy="0 ${camera_pitch} 0" xyz="${camera_x} ${camera_y} ${camera_z}"/>
</joint>

<!-- Scanner from camera -->
<!-- Origin is on the ground plane directly below the camera -->
<link name="scanner">
</link>
<joint name="scanner_origin" type="fixed">
  <parent link="camera"/>
  <child link="scanner"/>
  <origin rpy="0 ${-camera_pitch} 0" xyz="0 0 ${-camera_z}"/>
</joint>

</robot>
